src/map/utils/charutils.cpp

```diff
@@ -4918,13 +4947,130 @@ namespace charutils
+            PChar->pushPacket(new CRoeSparkUpdatePacket(PChar));
+    }
+
+    void SetEminenceRecordCompletion(CCharEntity* PChar, uint16 recordID, bool newStatus)
+    {
+        uint8 page = recordID / 8;
+        uint8 bit = recordID % 8;
+        if(newStatus)
+            PChar->m_eminenceLog.complete[page] |= (1 << bit);
+        else
+            PChar->m_eminenceLog.complete[page] &= ~(1 << bit);
+
+        for(int i = 0; i < 4; i++)
+        {
+            PChar->pushPacket(new CRoeQuestLogPacket(PChar, i));
+        }
+        charutils::SaveEminenceData(PChar);
+    }
+
+    bool GetEminenceRecordCompletion(CCharEntity* PChar, uint16 recordID)
+    {
+        uint8 page = recordID / 8;
+        uint8 bit = recordID % 8;
+        return PChar->m_eminenceLog.complete[page] & (1 << bit);
+    }
+
+    bool AddEminenceRecord(CCharEntity* PChar, uint16 recordID)
+    {
+        // TODO: Time limited records aren't implemented yet and can't be accepted normally.
+        //       For now we are refusing their IDs outright and protecting its slot from use here.
+        if(recordID > 2047)
+            return false;
+
+        // Per above, this i < 30 is correct.
+        for(int i = 0; i < 30; i++)
+        {
+            if(PChar->m_eminenceLog.active[i] == 0)
+            {
+                PChar->m_eminenceLog.active[i] = recordID;
+
+                PChar->pushPacket(new CRoeUpdatePacket(PChar));
+                PChar->pushPacket(new CMessageBasicPacket(PChar, PChar, recordID, 0, MSGBASIC_ROE_START));
+                SaveEminenceData(PChar);
+                return true;
+            }
+            else if(PChar->m_eminenceLog.active[i] == recordID)
+            {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    bool DelEminenceRecord(CCharEntity* PChar, uint16 recordID)
+    {
+        for(int i = 0; i < 31; i++)
+        {
+            if(PChar->m_eminenceLog.active[i] == recordID)
+            {
+                PChar->m_eminenceLog.active[i] = 0;
+                PChar->m_eminenceLog.progress[i] = 0;
+                // Shift entries up so records are shown in retail-accurate order.
+                for(int j = i; j < 29 && PChar->m_eminenceLog.active[j+1] != 0; j++)
+                {
+                    std::swap(PChar->m_eminenceLog.active[j], PChar->m_eminenceLog.active[j+1]);
+                    std::swap(PChar->m_eminenceLog.progress[j], PChar->m_eminenceLog.progress[j+1]);
+                }
+                PChar->pushPacket(new CRoeUpdatePacket(PChar));
+                charutils::SaveEminenceData(PChar);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool HasEminenceRecord(CCharEntity* PChar, uint16 recordID)
+    {
+        for(int i = 0; i < 31; i++)
+        {
+            if(PChar->m_eminenceLog.active[i] == recordID)
+            {
+                return true;
+            }
+        }
+        return false;
```

Comment:
I haven't dug into the logic for this, and it isn't that brutal of a loop to optimise, but: 
Could active quests be put into a hashmap and retrieved via lookup instead of search?

Reply:
+1 This is one area I've already optimized and will me pushing a commit for later. I want HasEminenceRecord to be basically-free since it will be called innumerable times once the check systems are written. In doing performance tests, a hashmap was actually about the same speed as this loop. I chose to implement a std::bitset of the player's active trials instead, which checking is about 6 times faster than this loop or an unordered_map.